tests/PHPStan/Analyser/data/bug-7094.php

```diff
@@ -0,0 +1,79 @@
+<?php declare(strict_types = 1);
+
+namespace Bug7094;
+
+/**
+ * @template T of array<string, mixed>
+ */
+trait AttributeTrait
+{
+    /**
+     * @param key-of<T> $key
+     */
+    public function hasAttribute(string $key): bool
+    {
+        $attr = $this->getAttributes();
+
+        return isset($attr[$key]);
+    }
+
+    /**
+     * @template K of key-of<T>
+     * @param K $key
+     * @param T[K] $val
+     */
+    public function setAttribute(string $key, $val): void
+    {
+        $attr = $this->getAttributes();
+        $attr[$key] = $val;
+        $this->setAttributes($attr);
```

Comment:
If I can bug you with a question @rvanvelzen - I am trying to implement this pretty much got the same setup as in this test file, but I get a weird result here:

I have this in setAttribute:

```
        $attr = $this->getAttributes() ?? [];
        \PHPStan\dumpType($attr);
        \PHPStan\dumpType($key);
        $attr[$key] = $val;
        \PHPStan\dumpType($attr);
        $this->setAttributes($attr);
```

Producing this output:

```
  28     Dumped type: array{language?: string, timezone?: string}
  29     Dumped type: string
  31     Dumped type: non-empty-array<string, array<int, string>|string>
  32     Parameter #1 $attributes of method User::setAttributes() expects array{language?: string, timezone?: string}, non-empty-array<string, array<int, string>|string> given.
```

You can see that $key got resolved to `string` which is too wide and then it widens the $attr type and breaks the setAttributes call.

I verified though with a `dumpType` of `$key` in `unsetAttribute` the direct `key-of<T>` works fine:

```
  52     Dumped type: 'language'|'timezone'
```

So somehow the indirection with the following seems to break it?

```
     * @template K of key-of<T>
     * @param K $key
```

Any idea what could be the cause? Why is this not breaking in the test suite? If I switch to use `@param key-of<T> $key` then as you described above it cannot resolve $val correctly as K is not set anymore and it ends up failing as well because it assumes every array key can be a union of any value present in the array.

Reply:
I don't have time to look into the root cause right now, but I believe using `K&key-of<T>` as the parameter type should already improve the situation.

I didn't actually spend much time looking at the parameter types inside the methods, that definitely can use some polishing.

Reply:
I tried that but could not make it work 100%, continuing this in https://github.com/phpstan/phpstan/issues/7380 so it does not get lost.